src/main/scala/beam/agentsim/infrastructure/power/SitePowerManager.scala

```diff
@@ -6,144 +6,211 @@ import beam.agentsim.infrastructure.ChargingNetwork.{ChargingCycle, ChargingStat
   protected var powerLimits = Map.empty[ChargingStation, PowerInKW]
   protected var powerCommands = Map.empty[Id[BeamVehicle], PowerInKW]
 
-  private[power] lazy val beamFederateMap: List[(Id[TAZ], List[ChargingStation], BeamFederate)] =
-    spmcConfigMaybe match {
-      case Some(spmcConfig) if spmcConfig.connect =>
-        logger.warn("ChargingNetworkManager should connect to a site power controller via Helics...")
-        Try {
-          // the same should be in 'all_taz' in src/main/python/gemini/site_power_controller_federate.py:311
-          // val selectedTazs = (920 until 930).map(_.toString).toSet
-          val tazIdToChargingStations = chargingNetworkHelper.allChargingStations.groupBy(_.zone.tazId)
-          //.filter { case (tazId, _) => selectedTazs.contains(tazId.toString) }
-          val numTAZs = tazIdToChargingStations.size
-          logger.info(s"Init SitePowerManager Federates for $numTAZs TAZes...")
-          val fedInfo = createFedInfo(
-            spmcConfig.coreType,
-            s"--federates=$numTAZs --broker_address=${spmcConfig.brokerAddress}",
-            spmcConfig.timeDeltaProperty,
-            spmcConfig.intLogLevel
-          )
-          tazIdToChargingStations.map { case (tazId, stations) =>
-            val fedName = spmcConfig.federatesPrefix + tazId.toString
-            logger.debug("getting federate {}", fedName)
-            val federate = getFederate(
-              fedName, //FED_BEAM_1
-              fedInfo,
-              spmcConfig.bufferSize,
-              beamServices.beamConfig.beam.agentsim.chargingNetworkManager.timeStepInSeconds,
-              Some(spmcConfig.federatesPublication),
-              if (spmcConfig.expectFeedback)
-                Some(spmcConfig.spmFederatesPrefix + tazId.toString + "/" + spmcConfig.spmSubscription)
-              else
-                None
-            )
-            logger.debug("Got federate for taz {}", tazId.toString)
-            (tazId, stations, federate)
-          }.seq
-        }.map { federates =>
-          logger.info("Initialized {} federates, now they are going to execution mode", federates.size)
-          val beamFederates = federates.map { case (_, _, beamFederate: BeamFederate) => beamFederate }.toSeq
-          enterExecutionMode(1.hour, beamFederates: _*)
-          logger.info("Entered execution mode")
-          federates.toList
-        }.recoverWith { case e =>
-          logger.warn(
-            s"Cannot initialize BeamFederate: ${e.getMessage}. " +
-            "ChargingNetworkManager is not connected to the SPMC"
-          )
-          Failure(e)
-        }.getOrElse(List.empty[(Id[TAZ], List[ChargingStation], BeamFederate)])
-      case _ => List.empty[(Id[TAZ], List[ChargingStation], BeamFederate)]
+  protected val siteMap: Map[Id[ParkingZoneId], Id[TAZ]] =
+    chargingNetworkHelper.allChargingStations
+      .groupBy(x => (x.zone.tazId, x.zone.parkingZoneId))
+      .keys
```

Comment:
Maybe use `map... distinct` instead of `groupBy.. keys`

Reply:
it is removed already